libraries/stdlib/jvm/src/kotlin/io/files/FileTreeWalk.kt

```diff
@@ -108,11 +137,40 @@ public class FileTreeWalk private constructor(
                 } else {
                     // Proceed to a sub-directory
                     state.push(directoryState(file))
-                    return gotoNext()
+                    return gotoNextDepthFirst()
                 }
             }
         }
 
+        /**
+         * Walking with BFS algorithm
+         */
+        private fun gotoNextBreadthFirst(): File? {
+            val walkState = state.poll() ?: return null
+            val file = walkState.root
+            if (currentDepth >= maxDepth) {
+                return file
+            }
+
+            // To process maximum depth
+            val nameCount = file.toComponents().size
+            if (nameCount > previousDepth) {
+                previousDepth = nameCount
+                currentDepth++
+            }
+            val listFiles = file.listFiles()
+            if (listFiles != null) {
+                for (fileItem in listFiles) {
+                    if (fileItem.isDirectory) {
+                        state.add(directoryState(fileItem))
+                    } else {
+                        state.add(SingleFileState(fileItem))
```

Comment:
How listing files from the root of this `SingleFileState` is prevented later?

Reply:
Hmm, I need to think about it.